diesel/src/connection/instrumentation.rs

```diff
@@ -313,3 +316,86 @@ where
         }
     }
 }
+
+#[diesel_derives::__diesel_public_if(
+    feature = "i-implement-a-third-party-backend-and-opt-into-breaking-changes"
+)]
+/// An optional dyn instrumentation.
+/// For ease of use, this type implements [`Deref`] and [`DerefMut`] to `&dyn Instrumentation`,
+/// falling back to a no-op implementation if no instrumentation is set.
+pub(crate) struct DynInstrumentation {
```

Comment:
Can we add the following part of your PR description as comment here?

> The DynInstrumentation type is useful because without it we actually did tend to return &mut Option<Box<dyn Instrumentation>> as &mut dyn Instrumentation from connection.instrumentation(), so downcasting would have to be done in these two steps by the user, which is counter-intuitive and doesn't seem ideal inside diesel itself either (because it induces extra dyn dispatch levels).

That provides the necessary context on why we use this abstraction instead of just `Option<Box<dyn Instrumentation>>` as before.

Also how about naming this type `OptionalInstrumentation` instead?

Reply:
> Also how about naming this type OptionalInstrumentation instead?

I have no strong preference for either naming. Just so that the reasoning for the current naming is brought to the table: initially I named it `OptionalDynInstrumentation` for completeness but then I thought it was pretty verbose and maybe unnecessarily so, considering that fundamentally, no instrumentation is also an instrumentation, so I opted to keep in the name the part that would let readers know "hey, this is essentially `Box<dyn Instrumentation>` for all intents and purposes".

Reply:
That sounds reasonable :+1: